### PR TITLE
cannon: Refactor instrumented_state.go to extract reusable helpers

### PR DESCRIPTION
**Pull request by mbaxter**
**Description**

Refactor `instrumented_state.go` to extract reusable helpers that will be useful for MTCannon. 

**Tests**

This is just refactor, so existing tests should provide coverage. 

**Metadata**

Part of: https://github.com/ethereum-optimism/optimism/issues/11642

### CODE REVIEW

File: cannon/mipsevm/exec/memory.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+	"fmt"
+
+	"github.com/ethereum-optimism/optimism/cannon/mipsevm/memory"
+)
+
+type MemTracker interface {
+	TrackMemAccess(addr uint32)
+}
+
+type MemoryTrackerImpl struct {
```
Comment (Inphi) [inline cannon/mipsevm/exec/memory.go:13]: Will there be another implementation of `MemTracker` for `multithreaded`? If not, perhaps it'll be better to use the implementation directly?

**Review by Inphi — APPROVED** [review]

File: cannon/mipsevm/exec/memory.go (hunk 1 above)
Comment (mbaxter) [reply cannon/mipsevm/exec/memory.go:13]: The only reason there's a separate interface `MemTracker` is that this is the interface needed by `mips_instructions` and `mips_syscalls`.  So having that just narrows down what the mips helpers can do.  The other methods available on `MemoryTrackerImpl ` are used by `InstrumentedState`.  Given that context, what are your thoughts on the best way to organize?

Comment (mbaxter) [reply cannon/mipsevm/exec/memory.go:13]: If you have a strong opinion, let me know.  Resolving and merging for now, but can fix this up later.

Comment (Inphi) [reply cannon/mipsevm/exec/memory.go:13]: Thanks for explaining. Looks good.

**Pull request merged**